rosbag2_cpp/include/rosbag2_cpp/writers/buffer_layer.hpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#define ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "rosbag2_storage/storage_interfaces/read_write_interface.hpp"
+#include "rosbag2_storage/storage_options.hpp"
+#include "rosbag2_cpp/visibility_control.hpp"
+
+namespace rosbag2_cpp
+{
+namespace writers
+{
+
+class ROSBAG2_CPP_PUBLIC BufferLayer
```

Comment:
bikeshedding here, but could we rename this to `CircularBuffer` or something alike? What's the `Layer` leading to here?

Reply:
I think that the idea was that this is the layer between writer and storage that is responsible for buffering (buffers themselves plus threading), thus the name. Is that correct, @pijaro?

If we separate threading from data structure:

CircularBuffer would be misleading, since this data structure is not a circular buffer.
Perhaps: CacheBuffer or MemoryConstrainedBuffer?

Reply:
Please take a look at current names of classes and let me know if they are ok